File: components/sdk/dotnet/src/Spas.Sdk.Metadata/Builders/ServiceIdentityBuilder.cs
using Spas.Sdk.Metadata.Models;

namespace Spas.Sdk.Metadata.Builders;

/// <summary>
/// Builder for service identity metadata.
/// </summary>
public class ServiceIdentityBuilder
{
    private string? _id;
    private string? _name;
    private string? _version;
    private string? _description;
    private string? _boundedContext;
    private readonly List<string> _capabilities = new();

    /// <summary>
    /// Sets the service ID (kebab-case identifier).
    /// </summary>
    public ServiceIdentityBuilder WithId(string id)
    {
        _id = id;
        return this;
    }

    /// <summary>
    /// Sets the service name.
    /// </summary>
    public ServiceIdentityBuilder WithName(string name)
    {
        _name = name;
        return this;
    }

    /// <summary>
    /// Sets the service version.
    /// </summary>
    public ServiceIdentityBuilder WithVersion(string version)
    {
        _version = version;
        return this;
    }

    /// <summary>
    /// Sets the service description.
    /// </summary>
    public ServiceIdentityBuilder WithDescription(string description)
    {
        _description = description;
        return this;
    }

    /// <summary>
    /// Sets the bounded context.
    /// </summary>
    public ServiceIdentityBuilder WithBoundedContext(string boundedContext)
    {
        _boundedContext = boundedContext;
        return this;
    }

    /// <summary>
    /// Adds a capability.
    /// </summary>
    public ServiceIdentityBuilder AddCapability(string capability)
    {
        _capabilities.Add(capability);
        return this;
    }

    /// <summary>
    /// Builds the service identity.
    /// </summary>
    public ServiceIdentity Build()
    {
        if (string.IsNullOrEmpty(_id))
        {
            throw new InvalidOperationException("Service id is required.");
        }

        if (string.IsNullOrEmpty(_name))
        {
            throw new InvalidOperationException("Service name is required.");
        }

        if (string.IsNullOrEmpty(_version))
        {
            throw new InvalidOperationException("Service version is required.");
        }

        if (string.IsNullOrEmpty(_boundedContext))
        {
            throw new InvalidOperationException("Bounded context is required.");
        }

        return new ServiceIdentity
        {
            Id = _id,
            Name = _name,
            Version = _version,
            Description = _description,
            BoundedContext = _boundedContext,
            Capabilities = _capabilities
        };
    }
}

File: components/sdk/java/spas-sdk-metadata/src/main/java/io/spas/sdk/metadata/builders/ServiceIdentityBuilder.java
package io.spas.sdk.metadata.builders;

import java.util.ArrayList;
import java.util.List;

/**
 * Fluent builder for service identity metadata.
 */
public final class ServiceIdentityBuilder {
    private String id;
    private String name;
    private String version;
    private String boundedContext;
    private String description;
    private final List<String> capabilities = new ArrayList<>();
    private String license;
    
    private ServiceIdentityBuilder() {}
    
    public static ServiceIdentityBuilder create() {
        return new ServiceIdentityBuilder();
    }
    
    public ServiceIdentityBuilder withId(String id) {
        this.id = id;
        return this;
    }
    
    public ServiceIdentityBuilder withName(String name) {
        this.name = name;
        return this;
    }
    
    public ServiceIdentityBuilder withVersion(String version) {
        this.version = version;
        return this;
    }
    
    public ServiceIdentityBuilder withBoundedContext(String boundedContext) {
        this.boundedContext = boundedContext;
        return this;
    }
    
    public ServiceIdentityBuilder withDescription(String description) {
        this.description = description;
        return this;
    }
    
    /**
     * Adds a capability to the service identity.
     * 
     * @param capability the capability identifier (e.g., "order-management")
     * @return this builder for method chaining
     * @deprecated Capabilities are automatically discovered from the {@code capabilities}
     *             attribute in {@link io.spas.sdk.metadata.annotations.SpasService} annotation.
     *             Manual capability registration is no longer necessary.
     *             This method will be removed in version 2.0.0.
     * @see io.spas.sdk.metadata.annotations.SpasService
     */
    @Deprecated(since = "1.1.0", forRemoval = true)
    public ServiceIdentityBuilder addCapability(String capability) {
        this.capabilities.add(capability);
        return this;
    }
    
    public ServiceIdentityBuilder withLicense(String license) {
        this.license = license;
        return this;
    }
    
    public ServiceIdentity build() {
        if (id == null || id.isEmpty()) {
            throw new IllegalArgumentException("id is required");
        }
        if (name == null || name.isEmpty()) {
            throw new IllegalArgumentException("name is required");
        }
        if (version == null || version.isEmpty()) {
            throw new IllegalArgumentException("version is required");
        }
        if (boundedContext == null || boundedContext.isEmpty()) {
            throw new IllegalArgumentException("boundedContext is required");
        }
        
        String normalizedDescription = (description == null || description.isBlank()) ? null : description;
        String normalizedLicense = (license == null || license.isBlank()) ? null : license;

        return new ServiceIdentity(
            id,
            name,
            version,
            boundedContext,
            normalizedDescription,
            capabilities.isEmpty() ? null : new ArrayList<>(capabilities),
            normalizedLicense
        );
    }
    
    /**
     * Simple record to hold service identity data.
     */
    public record ServiceIdentity(
        String id,
        String name,
        String version,
        String boundedContext,
        String description,
        List<String> capabilities,
        String license
    ) {}
}
